django/core/management/commands/shell.py

```diff
@@ -84,13 +84,13 @@ def python(self, options):
     def handle(self, **options):
         # Execute the command and exit.
         if options['command']:
-            exec(options['command'])
+            exec(options['command'], {})
```

Comment:
I think we should use `globals()`
```suggestion
            exec(options['command'], globals())
```
some commands will crash with an empty globals, e.g.
```
>>> python -m django shell -c <<EOF "
def f():
    with open(__file__, 'r'):
        pass
f()"
EOF
...
  File "/django/django/core/management/commands/shell.py", line 88, in handle
    exec(options['command'], {})
  File "<string>", line 5, in <module>
  File "<string>", line 3, in f
NameError: name '__file__' is not defined
```

Reply:
I made the change (890b1e35f2bfeafb7835de13da29a478b960fe72) as requested. I think it'll minimize the amount of backward compatibility breakage.

However I **did not add your test** case above. The **first** reason is that [`__file__`] is optional, never guaranteed to exist:

```console
$ python -c'print(__file__)'                                                                                                                                                                         
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name '__file__' is not defined
```

(I know Django generally assumes [`__file__`] exists, but it's [important](https://code.djangoproject.com/ticket/30950) to my use case that Django not become _more_ dependent on [`__file__`].)

Before 890b1e35f2bfeafb7835de13da29a478b960fe72, when we were just passing `{}`, [`__builtins__`] was the only member of the global namespace available to the code passed to `python -m django shell -c`. Other attributes [normally found](https://docs.python.org/3/reference/datamodel.html#special-method-names) in the [`__main__`] module (or other modules) are available by looking them up from [`__builtins__`] (ostensibly only in CPython; I have not tested on Pypy), making it appear (incorrectly) that we're in the [`builtins`](https://docs.python.org/3/library/builtins.html#module-builtins) module's namespace.

```console
$ python -m django shell -c'print( __name__, globals().keys(), sep="\n")'
builtins
dict_keys(['__builtins__']) 
```

Now that we're passing [`globals()`] to [`exec()`], it appears (correctly) that we're in the `django.core.management.commands.shell` module's namespace.

```console
$ python -m django shell -c'print(__name__); from pprint import pprint; pprint(list(globals().keys()))'
django.core.management.commands.shell
['__name__',
 '__doc__',
 '__package__',
 '__loader__',
 '__spec__',
 '__file__',
 '__cached__',
 '__builtins__',
 'os',
 'select',
 'sys',
 'traceback',
 'BaseCommand',
 'CommandError',
 'OrderedSet',
 'Command',
 'pprint']
```

This brings me to the **second** reason for why I didn't add your test case: It seems incorrect for `python -m django shell -c` to be guaranteed to run in the `django.core.management.commands.shell` module's namespace. To the extent that users have any expectation about the runtime environment of [`-c`](https://docs.python.org/3/using/cmdline.html#cmdoption-c), it's that it runs in the [`__main__`] module's namespace:

```console
$ python -c 'print(__name__)'                                                                                                                                             
__main__
```
This is neither important nor easy to emulate, so I'm fine using [`globals()`] to have the `python -m django shell -c` code run in `shell`'s namespace.

[`globals()`]: https://docs.python.org/3.9/library/functions.html#globals
[`exec()`]: https://docs.python.org/3.9/library/functions.html#exec
[`__file__`]: https://docs.python.org/3.9/reference/import.html#__file__
[`__builtins__`]: https://docs.python.org/3.9/reference/executionmodel.html#builtins-and-restricted-execution
[`__main__`]: https://docs.python.org/3.9/library/__main__.html

Reply:
Currently `shell` runs commands in the `django.core.management.commands.shell` scope and we should keep this behavior. IMO we should also add a test to avoid backward incompatibility changes in the future (which we almost introduce in this patch). I can add it in a separate commit.

Reply:
Let's add the test at the least to document the current behavior, this will protect us from unintended behavior change.